### PR TITLE
using DataFrame.resample with 'agg' method on non-existant columns provides unexpected behavior

### LINKED ISSUE #16766: using DataFrame.resample with 'agg' method on non-existant columns provides unexpected behavior
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from datetime import datetime

data = [
    {
        't': datetime(2017,6,1,0),
        'x': 1.0,
        'y': 2.0
    },
    {
        't': datetime(2017,6,1,1),
        'x': 2.0,
        'y': 2.0
    },
    {
        't': datetime(2017,6,1,2),
        'x': 3.0,
        'y': 1.5
    }

]

df = pd.DataFrame(data)
df = df.set_index('t')

# Perform a resample is get a binned time series DataFrame... this works fine
ts = df.resample('30T').agg({'x':['mean'],'y':['median']})
print ts['x'].shape
# (5,1)

# What if I put a field in there that doesn't exist? 
ts = df.resample('30T').agg({'x':['mean'],'y':['median'],'z':['sum']})
print ts['x'].shape
# (5,2) ??? I don't understand why the shape isn't (5,1)
print ts['x'].values
#[[ 1.   2. ]
# [ nan  nan]
# [ 2.   2. ]
# [ nan  nan]
# [ 3.   1.5]]
# Looks like a copy of the full aggregation even though I only requested the 'x' column
# Furthermore, now ts['z'] exists
```
Output:

(5, 1)
[[  1.]
 [ nan]
 [  2.]
 [ nan]
 [  3.]]
(5, 2)
[[ 1.   2. ]
 [ nan  nan]
 [ 2.   2. ]
 [ nan  nan]
 [ 3.   1.5]]

#### Problem description
I am using pandas on records from an Elasticsearch database. The queries are pulling from multiple indices with overlapping column/field names. Most records will have most of their data in common, but when available, I want to know the values of other fields. I'm essentially creating a time series for each column with a specific time-based binning and a per-column aggregation.

I think this should either ignore columns that don't exist or raise an exception. If it silently ignores these columns, ```ts['x']``` should return the same result as the first example. I spent several hours on a workaround today that required that I check for which columns were available for each aggregation and remove those from my agg dictionary. I feel like the current behavior doesn't have a purpose, but perhaps I'm missing something.

#### Expected Output
```python
(5, 1)
[[  1.]
 [ nan]
 [  2.]
 [ nan]
 [  3.]]
(5, 1)
[[  1.]
 [ nan]
 [  2.]
 [ nan]
 [  3.]]
```
OR
```python
ValueError("Column 'z' does not exist!")
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.5.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-514.2.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.20.2
pytest: None
pip: 9.0.1
setuptools: 0.9.8
Cython: None
numpy: 1.11.3
scipy: 0.18.1
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 1.5
pytz: 2012d
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by discort**
- [x] closes #16766
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/test_resample.py [hunk 1]
```diff
@@ -659,6 +659,21 @@ def f():
                                 'B': {'rb': ['mean', 'std']}})
             assert_frame_equal(result, expected, check_like=True)
 
+    def test_try_aggregate_non_existing_column(self):
```
Comment (discort) [inline pandas/tests/test_resample.py]: I'm not sure that this test is needed because `test_agg_misc` test covers this case. I added it to show reviewer that original issue is addressed. Let me know and I'll remove this new test.

File: doc/source/whatsnew/v0.23.0.txt [hunk 2]
```diff
@@ -646,6 +646,7 @@ Groupby/Resample/Rolling
 - Bug in :func:`DataFrame.groupby` where tuples were interpreted as lists of keys rather than as keys (:issue:`17979`, :issue:`18249`)
 - Bug in :func:`DataFrame.transform` where particular aggregation functions were being incorrectly cast to match the dtype(s) of the grouped data (:issue:`19200`)
 - Bug in :func:`DataFrame.groupby` passing the `on=` kwarg, and subsequently using ``.apply()`` (:issue:`17813`)
+- Bug in :func:`DataFrame.aggregate` on non-existant columns (:issue:`16766`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: Say what the issue was, also I think the issue was with `DataFrame.resample().aggregate`, not `DataFrame.aggregate`, right?

So something like

    Bug in :func:`DataFrame.resample` not raising a a `KeyError` when aggregating a non-existent column.

File: pandas/tests/test_resample.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/tests/test_resample.py]: Good to keep it.

**Comment by TomAugspurger** [comment]
Looks like your change is causing failures elsewhere. I'm not personally familiar with this code, I'm not sure what the best way forward is. Let me know if you need help though, and I can take a look.

**Comment by discort** [comment]
@TomAugspurger

**Comment by jreback** [comment]
thanks! keep em coming!

File: pandas/core/base.py [hunk 3]
```diff
@@ -392,6 +392,10 @@ def nested_renaming_depr(level=4):
 
                     elif isinstance(obj, ABCSeries):
                         nested_renaming_depr()
+                    elif isinstance(obj, ABCDataFrame) and \
+                            k not in obj.columns:
+                        raise ValueError(
```
Comment (jreback) [inline pandas/core/base.py]: actuallly I think this should be a KeyError. can you do another PR to fix?
